tslib/src/functions-throttled/ps-report-2-pupil-data/models.ts

```diff
@@ -0,0 +1,139 @@
+import moment from 'moment'
+
+export interface Pupil {
+  attendanceId: number | null
+  checkComplete: boolean
+  currentCheckId: number | null
+  dateOfBirth: moment.Moment
+  forename: string
+  gender: 'M' | 'F'
+  id: number
+  lastname: string
+  notTakingCheckReason: string | null
+  slug: string
+  schoolId: number
+  upn: string
+}
+
+export interface School {
+  estabCode: number
+  id: number
+  laCode: number
+  name: string
+  slug: string
+  urn: number
+}
+
+export interface CheckConfig {
+  audibleSounds: boolean
+  checkTime: number
+  colourContrast: boolean
+  fontSize: boolean
+  inputAssistance: boolean
+  loadingTime: number
+  nextBetweenQuestions: boolean
+  numpadRemoval: boolean
+  practice: boolean
+  questionReader: boolean
+  questionTime: number
+  compressCompletedCheck: boolean
+}
+
+interface Item {
+  f1: number
+  f2: number
+  questionNumber: number
+}
+
+export interface CheckForm {
+  id: number
+  items: Item[]
+  name: string
+}
+
+export interface Check {
+  id: number
+  checkCode: string
+  checkFormId: number
+  checkWindowId: number
+  complete: boolean
+  completedAt: moment.Moment
+  inputAssistantAddedRetrospectively: boolean
+  isLiveCheck: boolean
+  mark: number
+  processingFailed: boolean
+  pupilLoginDate: moment.Moment
+  received: boolean
+  restartNumber: number
+}
+
+export interface Device {
+  type: 'Tablet' | 'Desktop' | 'Mobile' | 'Other' | null
+  typeModel: string | null
+  browserFamily: string | null
+  browserMajorVersion: number | null
+  browserMinorVersion: number | null
+  browserPatchVersion: number | null
+  deviceId: string | null
+}
+
+export interface Input {
+  answerId: number
+  input: string
+  inputType: 'M' | 'K' | 'T' | 'P' | 'X'
+  browserTimestamp: moment.Moment
+}
+
+export type InputMap = Map<number, readonly Input[] | null>
+
+export interface Answer {
+  browserTimestamp: moment.Moment
+  id: number
+  inputs: readonly Input[] | null
+  isCorrect: boolean
+  question: string
+  questionCode: string
+  response: string
+}
+
+export interface Event {
+  browserTimestamp: moment.Moment
+  data?: object
+  id: number
+  isWarmup: boolean | null
+  question?: string | null
+  questionCode?: string | null
+  questionNumber?: number | null
+  type: string
+}
+
+export type CheckConfigOrNull = CheckConfig | null
```

Comment:
can you not express these with the optional operator `?`
or is it more the case that doing it like this makes it more explicit....

Reply:
Yes, we can express the same thing without the alias but these types are used in the code (mostly the data service) and I felt it was cleaner - as now all the types are defined in a single place making change easier!

This already proved useful during the refactor to make the output readonly.

Reply:
ok great.  makes sense, given that its easy to forget to add the operator everywhere you want to use the type